### PR TITLE
Add the guide on creating a new Ballerina module

### PR DESCRIPTION
**Pull request by ThisaruGuruge**
## Purpose

$Subject

### CODE REVIEW

File: docs/adding-a-new-ballerina-module.md [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+Enter selection (default: basic) [1..4]
+
+Select build script DSL:
+  1: Groovy
+  2: Kotlin
+Enter selection (default: Groovy) [1..2]
+
+Project name (default: module-ballerina-sample):
+
+Generate build using new APIs and behavior (some features may change in the next minor release)? (default: no) [yes, no]
```
Comment (NipunaRanasinghe) [inline docs/adding-a-new-ballerina-module.md:105]: Should we mention our recommended choices in here (maybe as inline comments)?

File: docs/adding-a-new-ballerina-module.md [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+
+This field is optional. It is only required if the version key cannot be inferred from the module name. (Eg.: `module-ballerina-jballerina.java.arrays` -> `stdlibJavaArraysVersion`, `module-ballerina-oauth2` -> `stdlibOAuth2Version`)
+
+>**Note:** The version key is case-sensitive and should be in the camel case format.
+
+Once the `module_list.json` file is updated, the [`Update Stdlib Dependency Graph`](https://github.com/ballerina-platform/ballerina-standard-library/actions/workflows/update_dependencies.yml) workflow will run automatically to update the Standard Library dashboard.
+
+Once updated this file, the module is ready to be released with the Ballerina Standard Library.
+
+### Adding the Module to Ballerina Daily Builds [Required]
```
Comment (NipunaRanasinghe) [inline docs/adding-a-new-ballerina-module.md:399]: Are we talking about the Full Build Pipeline in here? If so, better to mention it properly as in some repos we seem to have a daily build workflow in the repo itself as well. (e.g. https://github.com/ballerina-platform/module-ballerinax-kafka/actions/workflows/daily-build.yml)

File: docs/adding-a-new-ballerina-module.md [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+
+### The `spotbugs-exclude.xml` file [Optional]
+
+This file is used to skip specific spotbugs warnings/errors. This is required only if the module has Java (native) code and needs to skip specific spotbugs warnings/errors.
+
+## Adding a New Module
+
+After creating the module repository with the above structure, the following steps should be followed.
+
+### Adding the Module to the Ballerina Standard Library [Optional]
```
[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply docs/adding-a-new-ballerina-module.md:373]: This is not just to add to the dashboard, but to add to the release pipeline as well. So shall we keep it as it is?

[the remark this answers is not in the document]
Comment (NipunaRanasinghe) [reply docs/adding-a-new-ballerina-module.md:373]: I see.. +1 to keep it as is

File: docs/adding-a-new-ballerina-module.md [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+    6.10 [The `native` directory](#the-native-directory-gradle-submoduleoptional)
+    6.11 [Other Build Files](#other-build-files)
+        6.11.1 [The `LICENSE` file](#the-license-file-required)
+        6.11.2 [The `README.md` file](#the-readmemd-file-required)
+        6.11.3 [The `build.gradle` file](#the-buildgradle-file-required)
+        6.11.4 [The `changelog.md` file](#the-changelogmd-file-required)
+        6.11.5 [The `codecov.yml` file](#the-codecovyml-file-required)
+        6.11.6 [The `gradle.properties` file](#the-gradleproperties-file-required)
+        6.11.7 [The `gradlew` and `gradlew.bat` files](#the-gradlew-and-gradlewbat-files-required)
+        6.11.8 [The `settings.gradle` file](#the-settingsgradle-file-required)
```
[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply docs/adding-a-new-ballerina-module.md]: Fixed with e4bda59f235ace3aa80eb988f91b282b26dbfcc1

**Review by NipunaRanasinghe — DISMISSED** [review]
LGTM

File: docs/adding-a-new-ballerina-module.md [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+
+This workflow script is used to check the stale pull requests. It will add a label to the pull request if it is stale. This workflow script is triggered automatically on a schedule.
+
+##### The `trigger-load-tests.yml` workflow script [Optional]
+
+This workflow script is used to trigger the load tests of a module. It is needed only if the module has load tests. This workflow script is triggered automatically on a schedule and can be triggered manually as well.
+
+##### The `trivy-scan.yml` workflow script [Required]
+
+This workflow script is used to scan the module for vulnerabilities using Trivy. This workflow script is triggered automatically on a schedule and can be triggered manually as well. The status of this workflow is displayed on the Standard Library dashboard under the `Security Check` column.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:216]:
```suggestion
This workflow script is used to scan the module for vulnerabilities using Trivy. This workflow script is triggered automatically on a schedule and can be triggered manually as well. The status of this workflow is displayed on the Standard Library Dashboard under the `Security Check` column.
```

File: docs/adding-a-new-ballerina-module.md [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+
+##### The `update-spec.yml` workflow script [Optional]
+
+This workflow script is used to update the module specification in the Ballerina website. It is needed only if the module has a specification listed in the Ballerina website. This workflow script is triggered automatically when the `docs/spec` directory is updated.
+
+### The `ballerina` directory [Gradle Submodule][Required]
+
+This directory contains the Ballerina module source code, including the `Ballerina.toml`, `Module.md`, and `Package.md` files and the tests.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:224]:
```suggestion
This directory contains the Ballerina module source code including the `Ballerina.toml`, `Module.md`, and `Package.md` files and the tests.
```

File: docs/adding-a-new-ballerina-module.md [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+
+It also includes a `build.gradle` file which is used to build the Ballerina submodule. It uses the [Ballerina Gradle Plugin](https://github.com/ballerina-platform/plugin-gradle) to build the Ballerina module. The Ballerina Gradle plugin will add automated commits during the build to update the `Ballerina.toml`, `Dependencies.toml`, and `CompilerPlugin.toml` files.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:226]:
```suggestion
It also includes a `build.gradle` file, which is used to build the Ballerina submodule. It uses the [Ballerina Gradle plugin](https://github.com/ballerina-platform/plugin-gradle) to build the Ballerina module. The Ballerina Gradle plugin will add automated commits during the build to update the `Ballerina.toml`, `Dependencies.toml`, and `CompilerPlugin.toml` files.
```

File: docs/adding-a-new-ballerina-module.md [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+└── resources
+```
+
+#### The `checkstyle` directory [Optional]
+
+This directory contains the checkstyle configurations for the module. This is required only if a module has Java (native) code. It includes a `build.gradle` file.
+
+#### The `resources` directory [Required]
+
+This directory contains the resources required for the Ballerina module, including the `Ballerina.toml`, `BallerinaTest.toml`, and `CompilerPlugin.toml` files.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:244]:
```suggestion
This directory contains the resources required for the Ballerina module including the `Ballerina.toml`, `BallerinaTest.toml`, and `CompilerPlugin.toml` files.
```

File: docs/adding-a-new-ballerina-module.md [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+
+Each of these files have placeholders to replace the values during the build. The Ballerina Gradle plugin will update these files during the build and commit them as automated commits.
+
+>**Note:** The `CompilerPlugin.toml` file is required only if the module has a compiler plugin. The `BallerinaTest.toml` file is required only if the repository has `ballerina-tests` gradle submodule.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:248]:
```suggestion
>**Note:** The `CompilerPlugin.toml` file is required only if the module has a compiler plugin. The `BallerinaTest.toml` file is required only if the repository has the `ballerina-tests` Gradle submodule.
```

File: docs/adding-a-new-ballerina-module.md [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+
+### The `ballerina-tests` directory [Gradle Submodule][Optional]
+
+This directory contains the Ballerina tests that cannot be included in the `ballerina` directory. This includes the tests that are required to be run with the compiler plugin. It also includes a `build.gradle` file which is used to build and run the ballerina-tests submodule.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:252]:
```suggestion
This directory contains the Ballerina tests that cannot be included in the `ballerina` directory. This includes the tests that are required to be run with the compiler plugin. It also includes a `build.gradle` file, which is used to build and run the `ballerina-tests` submodule.
```

File: docs/adding-a-new-ballerina-module.md [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+
+`ballerina/http` and `ballerina/graphql` are example modules that have tests in the `ballerina-tests` directory.
+
+### The `compiler-plugin` directory [Gradle Submodule][Optional]
+
+This directory contains the compiler plugin source code. It also includes a `build.gradle` file which is used to build the compiler plugin submodule.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:258]:
```suggestion
This directory contains the compiler plugin source code. It also includes a `build.gradle` file, which is used to build the `compiler-plugin` submodule.
```

File: docs/adding-a-new-ballerina-module.md [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+
+### The `compiler-plugin-tests` directory [Gradle Submodule][Optional]
+
+This directory contains the compiler plugin tests written using [TestNG](https://testng.org/doc/). It also includes a `build.gradle` file which is used to build and run the compiler-plugin-tests submodule. This is required only if the module has a compiler plugin.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:262]:
```suggestion
This directory contains the compiler plugin tests written using [TestNG](https://testng.org/doc/). It also includes a `build.gradle` file, which is used to build and run the compiler-plugin-tests submodule. This is required only if the module has a compiler plugin.
```

File: docs/adding-a-new-ballerina-module.md [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+
+### The `docs` directory [Optional]
+
+This directory contains the module specifications. It includes a `proposals` directory to add the implemented proposals and the `spec.md` file that defines the module specification.
+
+### The `examples` directory [Gradle Submodule][Optional]
+
+This directory contains the examples for the module. It includes a `build.gradle` file which is used to build the examples submodule.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:270]:
```suggestion
This directory contains the examples for the module. It includes a `build.gradle` file, which is used to build the examples submodule.
```

Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:270]:
```suggestion
This directory contains the examples for the module. It includes a `build.gradle` file, which is used to build the `examples` submodule.
```

File: docs/adding-a-new-ballerina-module.md [hunk 14]
```diff
@@ -0,0 +1,487 @@
...
+
+This directory contains the deployment configurations for the load test including the `ingress.yaml` and `kustomization.yaml` files.
+
+#### The `results` directory [Required]
+
+This directory contains the results of the load test. The `trigger-load-tests` workflow script will upload the results to the `results` directory.
+
+#### The `scripts` directory [Required]
+
+This directory contains the scripts required for the load test. This includes the `run.sh` script which is used to run the load test.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:295]:
```suggestion
This directory contains the scripts required for the load test. This includes the `run.sh` script, which is used to run the load test.
```

File: docs/adding-a-new-ballerina-module.md [hunk 15]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `src` directory [Required]
+
+This directory contains the Ballerina source code of the load test.
+
+### The `native` directory [Gradle Submodule][Optional]
+
+This directory contains the Java native code of the module. This is required only if the module has native code. It also includes a `build.gradle` file which is used to build the native submodule.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:303]:
```suggestion
This directory contains the Java native code of the module. This is required only if the module has native code. It also includes a `build.gradle` file, which is used to build the `native` submodule.
```

File: docs/adding-a-new-ballerina-module.md [hunk 16]
```diff
@@ -0,0 +1,487 @@
...
+├── gradle.properties
+├── gradlew
+├── gradlew.bat
+├── settings.gradle
+├── spotbugs-exclude.xml
+```
+
+### The `LICENSE` file [Required]
+
+This file contains the license of the module. All the Ballerina Standard Library modules use `Apache-2` license.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:327]:
```suggestion
This file contains the license of the module. All the Ballerina Standard Library modules use the `Apache-2` license.
```

File: docs/adding-a-new-ballerina-module.md [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+
+### The `README.md` file [Required]
+
+This file contains the description and the build steps of the module. Usually the content of this file is similar to `Module.md` and `Package.md` files of the Ballerina module.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:331]:
```suggestion
This file contains the description and the build steps of the module. Usually, the content of this file is similar to the `Module.md` and `Package.md` files of the Ballerina module.
```

File: docs/adding-a-new-ballerina-module.md [hunk 18]
```diff
@@ -0,0 +1,487 @@
...
+
+This file contains the changelog of the module. The changelog should be updated with the changes of each feature/fix. Updating the changelog is added as a check in the pull request template to make sure each pull request has an entry in the changelog.
+
+### The `codecov.yml` file [Required]
+
+This file contains the Codecov configurations for the module. It is required to upload the Code coverage report for each pull request.
+
+### The `gradle.properties` file [Required]
+
+This file contains the dependency versions of the module. For Ballerina module dependencies, the naming convention of the version key is `ballerinaStdlib<moduleName>`. This convention should be followed strictly to make sure the automated dependency update workflows to work properly.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:349]:
```suggestion
This file contains the dependency versions of the module. For Ballerina module dependencies, the naming convention of the version key is `ballerinaStdlib<moduleName>`. This convention should be followed strictly for the automated dependency update workflows to work properly.
```

File: docs/adding-a-new-ballerina-module.md [hunk 19]
```diff
@@ -0,0 +1,487 @@
...
+
+### The `gradlew` and `gradlew.bat` files [Required]
+
+These files are auto-generated using the `gradle wrapper` command. These files are required to build the module using the Gradle wrapper.
+
+### The `settings.gradle` file [Required]
+
+This file contains the Gradle settings of the module. It includes the Gradle submodules of the module. The above mentioned directories related to `Gradle Submodule` should be added to this file using the following convention.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:357]:
```suggestion
This file contains the Gradle settings of the module. It includes the Gradle submodules of the module. The above-mentioned directories related to `Gradle Submodule` should be added to this file using the following convention.
```

File: docs/adding-a-new-ballerina-module.md [hunk 20]
```diff
@@ -0,0 +1,487 @@
...
+compiler-plugin-tests -> <module_name>-compiler-plugin-tests
+examples -> <module_name>-examples
+native -> <module_name>-native
+```
+
+This file also includes the Gradle plugins that are used to build the module.
+
+### The `spotbugs-exclude.xml` file [Optional]
+
+This file is used to skip specific spotbugs warnings/errors. This is required only if the module has Java (native) code and needs to skip specific spotbugs warnings/errors.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:372]:
```suggestion
This file is used to skip specific spotbugs warnings/errors. This is required only if the module has Java (native) code.
```

File: docs/adding-a-new-ballerina-module.md [hunk 21]
```diff
@@ -0,0 +1,487 @@
...
+
+## Adding a New Module
+
+After creating the module repository with the above structure, the following steps should be followed.
+
+### Adding the Module to the Ballerina Standard Library [Optional]
+
+This step is only required if the module is a part of the Ballerina Standard Library and it should be added to the Ballerina Standard Library dashboard and release process.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:380]:
```suggestion
This step is required only if the module is a part of the Ballerina Standard Library and if it should be added to the Ballerina Standard Library dashboard and release process.
```

File: docs/adding-a-new-ballerina-module.md [hunk 22]
```diff
@@ -0,0 +1,487 @@
...
+
+To add a module to the Ballerina Standard Library, add an entry in the [`module_list.json`](https://github.com/ballerina-platform/ballerina-standard-library/blob/main/release/resources/module_list.json) file in the [Ballerina Standard Library](https://github.com/ballerina-platform/ballerina-standrad-library) Repository.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:382]:
```suggestion
To add a module to the Ballerina Standard Library, add an entry in the [`module_list.json`](https://github.com/ballerina-platform/ballerina-standard-library/blob/main/release/resources/module_list.json) file in the [`ballerina-standard-library`](https://github.com/ballerina-platform/ballerina-standrad-library) repository.
```

File: docs/adding-a-new-ballerina-module.md [hunk 23]
```diff
@@ -0,0 +1,487 @@
...
+
+>**Note:** Do not edit the `stdlid_modules.json` file manually. It will be auto-generated once the `module_list.json` file is updated.
+
+The JSON entry has two fields.
+
+#### The `name` Field
+
+This field defines the name of the module. This should be the name of the repository. (Eg.: module-ballerina-io)
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:390]:
```suggestion
This field defines the name of the module. This should be the name of the repository (e.g., `module-ballerina-io`).
```

File: docs/adding-a-new-ballerina-module.md [hunk 24]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `version_key` Field
+
+The version key of the module. This is related to the version prefix mentioned in the [`gradle.properties`](#the-gradleproperties-file-required) file. (Eg.: ballerinaStdlibIo). It is used to add the module as a dependency to another repository including the [ballerina-distribution](https://github.com/ballerina-platform/ballerina-distribution).
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:394]:
```suggestion
The version key of the module. This is related to the version prefix mentioned in the [`gradle.properties`](#the-gradleproperties-file-required) file (e.g., `ballerinaStdlibIo`). It is used to add the module as a dependency to another repository including the [`ballerina-distribution`](https://github.com/ballerina-platform/ballerina-distribution).
```

File: docs/adding-a-new-ballerina-module.md [hunk 25]
```diff
@@ -0,0 +1,487 @@
...
+
+This field is optional. It is only required if the version key cannot be inferred from the module name. (Eg.: `module-ballerina-jballerina.java.arrays` -> `stdlibJavaArraysVersion`, `module-ballerina-oauth2` -> `stdlibOAuth2Version`)
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:396]:
```suggestion
This field is optional. It is only required if the version key cannot be inferred from the module name (e.g., `module-ballerina-jballerina.java.arrays` -> `stdlibJavaArraysVersion`, `module-ballerina-oauth2` -> `stdlibOAuth2Version`).
```

File: docs/adding-a-new-ballerina-module.md [hunk 26]
```diff
@@ -0,0 +1,487 @@
...
+
+>**Note:** The version key is case-sensitive and should be in the camel case format.
+
+Once the `module_list.json` file is updated, the [`Update Stdlib Dependency Graph`](https://github.com/ballerina-platform/ballerina-standard-library/actions/workflows/update_dependencies.yml) workflow will run automatically to update the Standard Library dashboard.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:400]:
```suggestion
Once the `module_list.json` file is updated, the [`Update Stdlib Dependency Graph`](https://github.com/ballerina-platform/ballerina-standard-library/actions/workflows/update_dependencies.yml) workflow will run automatically to update the Standard Library Dashboard.
```

File: docs/adding-a-new-ballerina-module.md [hunk 27]
```diff
@@ -0,0 +1,487 @@
...
+
+Once updated this file, the module is ready to be released with the Ballerina Standard Library.
+
+### Adding the Module to Ballerina Daily Full Build Pipeline [Required]
+
+This step is required to add a module to the Ballerina daily build process.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:406]:
```suggestion

```
This sentence seems redundant.

File: docs/adding-a-new-ballerina-module.md [hunk 28]
```diff
@@ -0,0 +1,487 @@
...
+
+To add a module to the Ballerina daily build process, add an entry in the [`module_list.json`](https://github.com/ballerina-platform/ballerina-release/blob/master/dependabot/resources/module_list.json) in the [Ballerina Release](https://github.com/ballerina-platform/ballerina-release) repository.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:408]:
```suggestion
To add a module to the Ballerina daily build process, add an entry in the [`module_list.json`](https://github.com/ballerina-platform/ballerina-release/blob/master/dependabot/resources/module_list.json) in the [`ballerina-release`](https://github.com/ballerina-platform/ballerina-release) repository.
```

File: docs/adding-a-new-ballerina-module.md [hunk 29]
```diff
@@ -0,0 +1,487 @@
...
+
+This Json file has two main fields.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:410]:
```suggestion
This JSON file has two main fields.
```

File: docs/adding-a-new-ballerina-module.md [hunk 30]
```diff
@@ -0,0 +1,487 @@
...
+
+- `standard_library` - This field contains an array of modules that are part of the Ballerina distribution. Add the module to this list if the module is released along with the Ballerina distribution.
+- `extended_library` - This field contains an array of modules that are not part of the Ballerina distribution. Add the module to this list if the module is an extended Ballerina module which is not released along with the Ballerina distribution.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:413]:
```suggestion
- `extended_library` - This field contains an array of modules that are not part of the Ballerina distribution. Add the module to this list if the module is an extended Ballerina module, which is not released along with the Ballerina distribution.
```

File: docs/adding-a-new-ballerina-module.md [hunk 31]
```diff
@@ -0,0 +1,487 @@
...
+
+Each entry in the above-mentioned arrays has the following fields.
+
+#### The `name` Field
+
+This field defines the name of the module. This should be the name of the repository. (Eg.: module-ballerina-io)
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:419]:
```suggestion
This field defines the name of the module. This should be the name of the repository (e.g., `module-ballerina-io`).
```

File: docs/adding-a-new-ballerina-module.md [hunk 32]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `version_key` Field
+
+The version key of the module. This is related to the version prefix mentioned in the [`gradle.properties`](#the-gradleproperties-file-required) file. (Eg.: ballerinaStdlibIo). It is used to add the module as a dependency to another repository including the [ballerina-distribution](https://github.com/ballerina-platform/ballerina-distribution) repository.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:423]:
```suggestion
The version key of the module. This is related to the version prefix mentioned in the [`gradle.properties`](#the-gradleproperties-file-required) file (e.g., `ballerinaStdlibIo`). It is used to add the module as a dependency to another repository including the [`ballerina-distribution`](https://github.com/ballerina-platform/ballerina-distribution) repository.
```

File: docs/adding-a-new-ballerina-module.md [hunk 33]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `level` Field
+
+This field denotes the level of the module in the dependency graph. This will be updated automatically. Do not update this field manually.
+
+#### The `group_id` Field
+
+This field defines the maven artifact group id of the module. The default value is `io.ballerina.stdlib`. If the group ID of the module is different, it should be updated in this field.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:431]:
```suggestion
This field defines the Maven artifact group ID of the module. The default value is `io.ballerina.stdlib`. If the group ID of the module is different, it should be updated in this field.
```

File: docs/adding-a-new-ballerina-module.md [hunk 34]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `artifact_id` Field
+
+This field defines the maven artifact id of the module. The default value is `<module_name>-ballerina`. If the artifact ID of the module is different, it should be updated in this field.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:435]:
```suggestion
This field defines the Maven artifact ID of the module. The default value is `<module_name>-ballerina`. If the artifact ID of the module is different, it should be updated in this field.
```

File: docs/adding-a-new-ballerina-module.md [hunk 35]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `default_branch` Field
+
+This field defines the default branch of the module. This field will be updated automatically. Do not update this field manually.
+
+#### The `auto_merge` Field
+
+This field defines whether to auto-merge the dependency update pull requests or not. Default value is `true`. Set this to `false` to stop auto-merging dependency update PRs (not recommended).
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:443]:
```suggestion
This field defines whether to auto-merge the dependency update pull requests or not. The default value is `true`. Set this to `false` to stop auto merging dependency update PRs (not recommended).
```

File: docs/adding-a-new-ballerina-module.md [hunk 36]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `push_to_central` Field
+
+This field is used to define whether the module is being pushed to the Ballerina central or not. Default value is `true`. Set this to `false` for non-central repositories such as Ballerina tools repositories (Eg.: `ballerina-openapi-tools`, `ballerina-graphql-tools`).
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:447]:
```suggestion
This field is used to define whether the module is being pushed to the Ballerina Central or not. The default value is `true`. Set this to `false` for non-central repositories such as Ballerina tools repositories (e.g.,`ballerina-openapi-tools`, `ballerina-graphql-tools`, etc.).
```

File: docs/adding-a-new-ballerina-module.md [hunk 37]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `is_extended_library` Field
+
+This field is used to define whether the module is an extended library or not. Default value is `true`. Set this to `false` for repositories that are released along with the [ballerina-distribution](https://github.com/ballerina-platform/ballerina-distribution).
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:451]:
```suggestion
This field is used to define whether the module is an extended library or not. The default value is `true`. Set this to `false` for repositories that are released along with the [`ballerina-distribution`](https://github.com/ballerina-platform/ballerina-distribution).
```

File: docs/adding-a-new-ballerina-module.md [hunk 38]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `build_action_file` Field
+
+This field is to define the default build action file of the module. This field will be updated automatically. Do not update this field manually.
+
+#### The `send_notification` Field
+
+This field is used to define whether to send a notification to the chat on build failures. Default value is `true`. Set this to `false` to stop sending notifications on build failures (not recommended).
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:459]:
```suggestion
This field is used to define whether to send a notification to the chat on build failures. The default value is `true`. Set this to `false` to stop sending notifications on build failures (not recommended).
```

File: docs/adding-a-new-ballerina-module.md [hunk 39]
```diff
@@ -0,0 +1,487 @@
...
+
+#### The `dependents` Field
+
+This field is used to add the dependents of the module. This field will be updated automatically. Do not update this field manually.
+
+Once the `module_list.json` is updated, the [Update Dependency Graph](https://github.com/ballerina-platform/ballerina-release/actions/workflows/update_dependency_graph.yml) workflow will run automatically to update the dependency graph.
+
+### Adding the Module to Ballerina Distribution [Optional]
+
+This is required only if the module is released with the [ballerina-distribution](https://github.com/ballerina-platform/ballerina-distribution).
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:469]:
```suggestion
This is required only if the module is released with the [`ballerina-distribution`](https://github.com/ballerina-platform/ballerina-distribution).
```

File: docs/adding-a-new-ballerina-module.md [hunk 40]
```diff
@@ -0,0 +1,487 @@
...
+
+To add the module to the Ballerina distribution, follow the below steps.
+
+#### Update the `gradle.properties` File [Required]
+
+The module version should be added to the [`gradle.properties`](https://github.com/ballerina-platform/ballerina-distribution/blob/master/gradle.properties) file in the [ballerina-distribution](https://github.com/ballerina-platform/ballerina-distribution).
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:475]:
```suggestion
The module version should be added to the [`gradle.properties`](https://github.com/ballerina-platform/ballerina-distribution/blob/master/gradle.properties) file in the [`ballerina-distribution`](https://github.com/ballerina-platform/ballerina-distribution).
```

File: docs/adding-a-new-ballerina-module.md [hunk 41]
```diff
@@ -0,0 +1,487 @@
...
+
+>**Note:** The timestamp version can be found in the module repository under the `packages` section.
+
+#### Update the `build.gradle` File [Required]
+
+The module should be added as a configuration in the [`build.gradle`](https://github.com/ballerina-platform/ballerina-distribution/blob/master/build.gradle) file in the [ballerina-distribution](https://github.com/ballerina-platform/ballerina-distribution).
+
+>**Note:** The module should be added under the `ballerinaStdlibs` configuration.
+
+That's it. The module is now added to the Ballerina daily builds and release processes.
```
Comment (praneesha) [inline docs/adding-a-new-ballerina-module.md:487]:
```suggestion
The module is now added to the Ballerina daily builds and release processes.
```

**Review by shafreenAnfar — APPROVED** [review]
LGTM!

File: docs/adding-a-new-ballerina-module.md [hunk 42]
```diff
@@ -0,0 +1,487 @@
...
+        * 6.11.8 [The `settings.gradle` file](#the-settingsgradle-file-required)
+        * 6.11.9 [The `spotbugs-exclude.xml` file](#the-spotbugs-excludexml-file-optional)
+7. [Adding a New Module](#adding-a-new-module)
+    * 7.1 [Adding the Module to the Ballerina Standard Library](#adding-the-module-to-the-ballerina-standard-library-optional)
+    * 7.2 [Adding the Module to Ballerina Daily Full Build Pipeline](#adding-the-module-to-ballerina-daily-full-build-pipeline-required)
+    * 7.3 [Adding the Module to Ballerina Distribution](#adding-the-module-to-ballerina-distribution-optional)
+
+## Introduction
+
+This guide helps to understand the file structure, build scripts, and workflow scripts of a Ballerina module. It also explains how to add a new module to the Ballerina daily build and release pipelines. Each directory/file has tags defined inside `[]` to indicate whether it is required or optional and relation to `ballerina` and `ballerinax` module.
```
[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply docs/adding-a-new-ballerina-module.md:52]: This is incorrect. Required or optional does not relate to the org name.

File: docs/adding-a-new-ballerina-module.md [hunk 43]
```diff
@@ -0,0 +1,487 @@
...
+
+##### The `process-load-test-result.yml` workflow script [Optional]
+
+This workflow script is used to process the load test results of a module. It is needed only if the module has load tests. The status of this workflow is displayed on the Standard Library dashboard under the `Load Test Results` column.
+
+>**Note:** In case of a failure of this workflow due to non-persistent issue, close the failed pull request, delete the branch that sent the pull request and re-run the failed workflow.
+
+##### The `publish-release.yml` workflow script [Required]
+
+This workflow script is used to release the module. It will create a new git tag, create a GitHub release, publish the artifacts to GitHub packages, publish the module to the Ballerina central, and finally sends a post-release sync pull request.
```
[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply docs/adding-a-new-ballerina-module.md:196]: This creates a new git tag, it's not a GitHub tag.

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

File: docs/adding-a-new-ballerina-module.md (hunk 42 above)
[the remark this answers is not in the document]
Comment (praneesha) [reply docs/adding-a-new-ballerina-module.md]: Use of the 'relation' word is not clear in the original context. Can we change it?

[the remark this answers is not in the document]
Comment (praneesha) [reply docs/adding-a-new-ballerina-module.md]: How about: `Each directory/file has tags defined inside `[]` to indicate whether it is required or optional and whether it is related to the `ballerina` or `ballerinax` modules.`

[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply docs/adding-a-new-ballerina-module.md]: +1

[the remark this answers is not in the document]
Comment (ThisaruGuruge) [reply docs/adding-a-new-ballerina-module.md]: On second thoughts, we can remove this sentence altogether. Removed with c5cfbeab9f683c2e970b91a691f22e2efb18d32e

**Review by NipunaRanasinghe — APPROVED** [review]

**Pull request merged**